File: notification-schema-java/src/main/java/uk/gov/defra/tracesx/notificationschema/representation/PartTwo.java
package uk.gov.defra.tracesx.notificationschema.representation;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import uk.gov.defra.tracesx.notificationschema.representation.enumeration.InspectionRequired;
import uk.gov.defra.tracesx.notificationschema.representation.serialisation.IsoOffsetDateTimeDeserializer;
import uk.gov.defra.tracesx.notificationschema.representation.serialisation.IsoOffsetDateTimeSerializer;
import uk.gov.defra.tracesx.notificationschema.validation.ValidationMessageCode;
import uk.gov.defra.tracesx.notificationschema.validation.annotations.CedOrCvedpControlledDestination;
import uk.gov.defra.tracesx.notificationschema.validation.annotations.ChedpIdentityCheck;
import uk.gov.defra.tracesx.notificationschema.validation.annotations.ChedpPhysicalCheck;
import uk.gov.defra.tracesx.notificationschema.validation.annotations.CvedaControlledDestination;
import uk.gov.defra.tracesx.notificationschema.validation.annotations.LaboratoryTestsNotAdded;
import uk.gov.defra.tracesx.notificationschema.validation.annotations.LaboratoryTestsPending;
import uk.gov.defra.tracesx.notificationschema.validation.groups.NotificationCedOrCvedpFieldValidation;
import uk.gov.defra.tracesx.notificationschema.validation.groups.NotificationCvedaFieldValidation;
import uk.gov.defra.tracesx.notificationschema.validation.groups.NotificationCvedpFieldValidation;
import uk.gov.defra.tracesx.notificationschema.validation.groups.NotificationHighRiskNonChedppFieldValidation;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Set;
import javax.validation.Valid;
import javax.validation.constraints.NotNull;

@Builder
@Data
@JsonInclude(Include.NON_EMPTY)
@NoArgsConstructor(access = AccessLevel.PUBLIC)
@AllArgsConstructor(access = AccessLevel.PUBLIC)
@LaboratoryTestsNotAdded(
    groups = NotificationHighRiskNonChedppFieldValidation.class,
    message =
        "{uk.gov.defra.tracesx.notificationschema.representation.parttwo"
            + ".reasonlaboratorytestsnotadded.not.empty}")
@LaboratoryTestsPending(
    groups = NotificationHighRiskNonChedppFieldValidation.class,
    message =
        "{uk.gov.defra.tracesx.notificationschema.representation.parttwo.laboratorytestspending"
            + ".not.empty}")
@CvedaControlledDestination(
    groups = NotificationCvedaFieldValidation.class,
    message =
        "{uk.gov.defra.tracesx.notificationschema.representation.parttwo.controlleddestination"
            + ".not.empty}")
@CedOrCvedpControlledDestination(
    groups = NotificationCedOrCvedpFieldValidation.class,
    message =
        "{uk.gov.defra.tracesx.notificationschema.representation.parttwo.controlleddestination"
            + ".not.empty}")
@ChedpIdentityCheck(groups = NotificationCvedpFieldValidation.class)
@ChedpPhysicalCheck(groups = NotificationCvedpFieldValidation.class)
public class PartTwo {

  @Valid
  private Decision decision;

  @Valid
  private ConsignmentCheck consignmentCheck;

  private ImpactOfTransportOnAnimals impactOfTransportOnAnimals;

  @NotNull(
      groups = NotificationHighRiskNonChedppFieldValidation.class,
      message =
          "{uk.gov.defra.tracesx.notificationschema.representation.parttwo"
              + ".laboratorytestsrequired.not.null}")
  private Boolean laboratoryTestsRequired;

  @Valid
  private LaboratoryTests laboratoryTests;

  private Boolean resealedContainersIncluded;

  private List<String> resealedContainers;

  private List<NotificationSealsContainers> resealedContainersMapping;

  @Valid
  private ControlAuthority controlAuthority;

  private String bipLocalReferenceNumber;
  private String signedOnBehalfOf;
  private String onwardTransportation;
  private Set<ValidationMessageCode> consignmentValidation;
  private EconomicOperator controlledDestination;

  @JsonSerialize(using = IsoOffsetDateTimeSerializer.class)
  @JsonDeserialize(using = IsoOffsetDateTimeDeserializer.class)
  private LocalDateTime checkDate;

  private List<AccompanyingDocument> accompanyingDocuments;

  private List<CommodityChecks> commodityChecks;

  private Boolean phsiAutoCleared;

  private Boolean hmiAutoCleared;

  private InspectionRequired inspectionRequired;

  private InspectionOverride inspectionOverride;
}

File: imports-frontend-entities/src/entities/part_two.js
const _ = require('lodash')
const handler = require('./base/handler')
const Decision = require('./decision')
const ConsignmentCheck = require('./consignment_check')
const ImpactOfTransportOnAnimals = require(
    './impact_of_transportation_on_animals')
const LaboratoryTests = require('./laboratory_tests')
const ControlAuthority = require('./control_authority')
const ConsignmentValidation = require('./consignment_validation')
const EconomicOperator = require('./economic_operator')
const AccompanyingDocument = require('./accompanying_document')
const CommodityChecks = require('./commodity_checks')
const InspectionOverride = require('./inspection_override')
const { getList } = require('../utils/list')

module.exports = class PartTwo {

  constructor(obj = {}) {

    this.decision = _.get(obj, 'decision') ? new Decision(obj.decision)
        : undefined
    this.consignmentCheck = _.get(obj, 'consignmentCheck')
        ? new ConsignmentCheck(obj.consignmentCheck) : undefined
    this.impactOfTransportOnAnimals = _.get(obj, 'impactOfTransportOnAnimals')
        ? new ImpactOfTransportOnAnimals(obj.impactOfTransportOnAnimals)
        : undefined
    this.laboratoryTestsRequired = obj.laboratoryTestsRequired
    this.laboratoryTests = _.get(obj, 'laboratoryTests') ? new LaboratoryTests(
        obj.laboratoryTests) : undefined
    this.resealedContainersIncluded = obj.resealedContainersIncluded
    this.resealedContainers = _.get(obj, 'resealedContainers', [])
    this.resealedContainersMapping = getSealsContainers(_.get(obj, 'resealedContainersMapping', []))
    this.controlAuthority = _.get(obj, 'controlAuthority')
        ? new ControlAuthority(obj.controlAuthority) : undefined
    this.bipLocalReferenceNumber = obj.bipLocalReferenceNumber
    this.signedOnBehalfOf = obj.signedOnBehalfOf
    this.onwardTransportation = obj.onwardTransportation
    this.consignmentValidation = getConsignmentValidation(
        _.get(obj, 'consignmentValidation', []))
    this.checkDate = obj.checkDate
    this.controlledDestination = _.get(obj, 'controlledDestination')
        ? new EconomicOperator(obj.controlledDestination) : undefined
    this.accompanyingDocuments = getList(_.get(obj, 'accompanyingDocuments', []), AccompanyingDocument)
    this.commodityChecks = getList(_.get(obj, 'commodityChecks', []), CommodityChecks)
    this.phsiAutoCleared = obj.phsiAutoCleared
    this.hmiAutoCleared = obj.hmiAutoCleared
    this.inspectionRequired = obj.inspectionRequired
    this.inspectionOverride = _.get(obj, 'inspectionOverride')
        ? new InspectionOverride(obj.inspectionOverride) : undefined
    this.autoClearedDateTime = obj.autoClearedDateTime

    return Object.seal(new Proxy(this, handler))
  }
}

const getSealsContainers = inList => {

  const sealsContainersList = []
  _.forEach(inList, seal => {
    sealsContainersList.push(new NotificationSealsContainers(seal))
})

  return sealsContainersList
}

const getConsignmentValidation = inList => {

  const consignmentValidationList = []
  _.forEach(inList, consignmentValidation => {
    consignmentValidationList.push(
        new ConsignmentValidation(consignmentValidation))
  })

  return consignmentValidationList
}